src/pygenstability/data_clustering.py

```diff
@@ -0,0 +1,202 @@
+        distance_threshold=np.inf,
+    ):
+        # parameters
+        self.metric = metric
+        self.method = method
+        self.k = k
+        self.delta = delta
+        self.distance_threshold = distance_threshold
+
+        # attributes
+        self.adjacency_ = None
+
+    def fit(self, X):
+        """Construct graph from samples-by-features matrix."""
+
+        # if precomputed take X as adjacency matrix
+        if self.method == "precomputed":
+            assert (
+                X.shape[0] == X.shape[1]
+            ), "Precomputed matrix should be a square matrix."
+            self.adjacency_ = X
+            return self.adjacency_
+
+        # compute normalised distance matrix
+        D = squareform(pdist(X, metric=self.metric))
+        D_norm = D / np.amax(D)
+
+        # compute normalised similarity matrix
+        S = 1 - D_norm
+
+        # sparsify distance matrix with CkNN or kNN method
+        if self.method == "cknn":
+            sparse = compute_CkNN(D_norm, self.k, self.delta)
+
+        elif self.method == "knn":
+            sparse = kneighbors_graph(
+                D_norm, n_neighbors=self.k, metric="precomputed"
+            ).toarray()
+
+        # undirected distance backbone is given by sparse graph and MST
+        mst = minimum_spanning_tree(D_norm)
+        backbone = np.array((mst + mst.T + sparse + sparse.T) > 0, dtype=int)
+
+        # apply distance threshold to backbone
+        backbone *= np.array(D <= self.distance_threshold, dtype=int)
+
+        # adjacency matrix has weights of similarity matrix
+        self.adjacency_ = S * backbone
+
+        return self.adjacency_
+
+
+class DataClustering(GraphConstruction):
+    """Data clustering."""
+
+    def __init__(
+        self,
+        metric="euclidean",
+        graph_method="cknn",
+        k=5,
+        delta=1.0,
+        distance_threshold=np.inf,
+        **pgs_kwargs,
+    ):
+
+        # initialise parameters for graph construction
+        super().__init__(
+            metric=metric,
+            method=graph_method,
+            k=k,
+            delta=delta,
+            distance_threshold=distance_threshold,
+        )
+
+        # store PyGenStability kwargs
+        self.pgs_kwargs = pgs_kwargs
+
+        # attributes
+        self.results_ = {}
+
+    @property
+    def labels_(self):
+        """Return labels for robust paritions."""
+        labels = []
+
+        assert (
+            "selected_partitions" in self.results_.keys()
+        ), "Run PyGenStability with optimal scale selection first."
+
+        # store labels of robust partitions
+        for i in self.results_["selected_partitions"]:
+
+            # only store non-trivial robust partitions
+            robust_partition = self.results_["community_id"][i]
+            if not np.allclose(robust_partition, np.zeros(self.adjacency_.shape[0])):
+                labels.append(robust_partition)
+
+        return labels
+
+    def fit(self, X):
```

Comment:
I would also not use fit as a name here, but maybe 'run' ? Its not quite fitting things, but running pgs, no?

Reply:
In sklearn fit() is also used when a clustering algorithm is applied, see here for example kmean: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html#sklearn.cluster.KMeans.fit So I think sticking to fit() might make it easier for users.

Reply:
Great! I've implemented that.